processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -514,31 +516,32 @@ public InputRow formatRow(InputRow row)
    *
    * @return the number of rows in the data set after adding the InputRow
    */
-  public int add(InputRow row) throws IndexSizeExceededException
+  public IncrementalIndexAddResult add(InputRow row) throws IndexSizeExceededException
   {
     return add(row, false);
   }
 
-  public int add(InputRow row, boolean skipMaxRowsInMemoryCheck) throws IndexSizeExceededException
+  public IncrementalIndexAddResult add(InputRow row, boolean skipMaxRowsInMemoryCheck) throws IndexSizeExceededException
   {
-    TimeAndDims key = toTimeAndDims(row);
-    final int rv = addToFacts(
+    Pair<TimeAndDims, List<String>> timeAndDimsResult = toTimeAndDims(row);
+    final Pair<Integer, List<String>> addToFactsResult = addToFacts(
         metrics,
         deserializeComplexMetrics,
         reportParseExceptions,
         row,
         numEntries,
-        key,
+        timeAndDimsResult.lhs,
         in,
         rowSupplier,
         skipMaxRowsInMemoryCheck
     );
     updateMaxIngestedTime(row.getTimestamp());
-    return rv;
+    ParseException parseException = getCombinedParseException(row, timeAndDimsResult.rhs, addToFactsResult.rhs);
+    return new IncrementalIndexAddResult(addToFactsResult.lhs, parseException);
   }
 
   @VisibleForTesting
-  TimeAndDims toTimeAndDims(InputRow row)
+  Pair<TimeAndDims, List<String>> toTimeAndDims(InputRow row)
```

Comment:
Same here.

Reply:
Changed this to use a new class instead of Pair